Introduce unified `DataSourceExec` for provided datasources, remove `ParquetExec`, `CsvExec`, etc

mertak-synnada opened the pull request:
## Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

- Closes #13838.
- Closes https://github.com/apache/datafusion/pull/14502
- Closes https://github.com/apache/datafusion/issues/14337

## Rationale for this change

This PR merges all Data sources into one Execution Plan, named DataSourceExec, and a single trait named DataSource which is inspired by DataSink. File-related shared behaviors are merged into FileSourceConfig, and a MemorySourceConfig is created. MemoryExec, ArrowExec, AvroExec, CsvExec, NdJsonExec, and ParquetExec are deprecated and merged into DataSourceExec.

## What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

mertak-synnada commented on datafusion/physical-plan/src/source.rs:
```diff
@@ -0,0 +1,201 @@
+
+/// Unified data source for file formats like JSON, CSV, AVRO, ARROW, PARQUET
+#[derive(Clone, Debug)]
+pub struct DataSourceExec {
```
`DataSourceExec` is responsible for calculating properties, managing cache, and using `source` to get source-related information and create a stream.

The repartition function is separated into two, DataSourceExec handles the part of `ExecutionPlan` and passes the rest to the source definition.

mertak-synnada commented on datafusion/physical-plan/src/memory.rs:
```diff
@@ -168,38 +137,41 @@ impl ExecutionPlan for MemoryExec {
 
+    /// Create a new `DataSourceExec` plan for reading in-memory record batches
+    /// The provided `schema` should not have the projection applied.
+    pub fn try_new_exec(
```
Same as FileSourceConfig::new_exec

alamb commented:
Thank you @mertak-synnada  -- I plan to study this PR carefully tomorrow

berkaysynnada commented:
I'll also be reviewing this in detail today.

alamb commented:
I ran out of time for this today, but I will do it tomorrow

alamb commented on datafusion/core/src/datasource/physical_plan/arrow_file.rs:
```diff
@@ -223,6 +226,65 @@ impl ExecutionPlan for ArrowExec {
+/// Arrow configuration struct that is given to DataSourceExec
+/// Does not hold anything special, since [`FileScanConfig`] is sufficient for arrow
+#[derive(Clone, Default)]
+pub struct ArrowConfig {
```
Since this implements `FileSource` I would personally suggest a name like `ArrowSource` (and `ParquetSource`, `AvroSource`, etc) for consistency

In my mind this is not really "config" it is the runtime state needed to read arrow files. But I think I could be conviced that Config is ok too.

alamb commented on datafusion/core/src/datasource/physical_plan/avro.rs:
```diff
@@ -18,24 +18,31 @@
 
 /// Execution plan for scanning Avro data source
 #[derive(Debug, Clone)]
+#[deprecated(since = "46.0.0", note = "use DataSourceExec instead")]
```
Thank you for deprecating this.

In order to avoid having to maintain 2 copies of the code (and avoid having them drift out of sync), I think we should consider leaving the interface the same but changing the internal implementation. 

So that would be something like:
```rust
pub struct AvroExec {
    inner: DataSourceExec, // wraps a DataSourceExec
}
```

alamb commented on datafusion/core/src/datasource/physical_plan/file_scan_config.rs:
```diff
@@ -156,6 +161,8 @@ impl FileScanConfig {
             limit: None,
             table_partition_cols: vec![],
             output_ordering: vec![],
+            file_compression_type: FileCompressionType::UNCOMPRESSED,
+            new_lines_in_values: false,
```
Why are these added to FileScanConfig? Is it because they are common for JSON and CSV or something?

I ask because I don't think they are relevant for AVRO and Parquet

mertak-synnada replied:
Yes, these are only common in CSV and JSON but since their defaults are not affecting I thought we could add them in the common area to avoid code duplication, but if it feels wrong for more possible file types, we can remove them from here

alamb commented on datafusion/core/src/datasource/physical_plan/parquet/mod.rs:
```diff
@@ -532,10 +342,648 @@ impl ParquetExec {
+/// [`RecordBatch`]: arrow::record_batch::RecordBatch
+/// [`SchemaAdapter`]: crate::datasource::schema_adapter::SchemaAdapter
+/// [`ParquetMetadata`]: parquet::file::metadata::ParquetMetaData
+#[derive(Clone, Default, Debug)]
```
if we are moving lots of code anyway, maybe this struct could go in its own module to break up the structure more

alamb reviewed (commented):
Thank you very much @mertak-synnada  . This is a pretty epic PR 👏 

I think the idea of refactoring out the common ExecutionPlan functionality for provided  formats like Parquet/Avro/CSV etc makes a lot of sense

I had a few questions on the exact choices of traits but at a high level I really like this structure and where it is headed

# Impact Analysis

My understanding of the impacts of this PR on downstream crates are:
1. Anyone with a TableProvider that returns a custom ExecutionPlan is unaffected
2. Anyone with a TableProvider / code that directly handles `ParquetExec` will need to update to use the new DataSourceExec

# Next steps

So next steps I think would be  test with delta-rs  as suggested on https://github.com/apache/datafusion/issues/13838#issuecomment-2554943551

Specifically it looks to me like they use ParquetExec a few places directly
https://github.com/search?q=repo%3Adelta-io%2Fdelta-rs%20ParquetExec&type=code

I will also see if we can find time internally at InfluxData to try updating our code to use this new API and report our findings.

andygrove commented:
@parthchandra @mbutrovich, could you take a look since this may impact our current work to move to ParquetExec

parthchandra commented:
> @parthchandra @mbutrovich, could you take a look since this may impact our current work to move to ParquetExec

It looks for the most part that we should be able to move to DataSourceExec. I'll take a second closer look to make sure that our schema adapter work is not affected too much.

jayzhan-synnada commented on datafusion/core/src/datasource/data_source.rs:
```diff
@@ -0,0 +1,264 @@
+}
+
+impl FileSourceConfig {
+    // TODO: This function should be moved into DataSourceExec once FileScanConfig and FileSourceConfig moved out of datafusion/core
```
`DataSourceExec` is in `physical-plan` not `core`.

Xuanwo commented on datafusion/core/src/datasource/data_source.rs:
```diff
@@ -0,0 +1,264 @@
+    /// Creates a `dyn FileOpener` based on given parameters
+    fn create_file_opener(
+        &self,
+        object_store: datafusion_common::Result<Arc<dyn ObjectStore>>,
```
Hi, thank you very much for creating this. It seems like a great time for us to consider the possibility of integrating with [Apache OpenDAL](https://github.com/apache/opendal): an Open Data Access Layer that enables seamless interaction with diverse storage services.

In this PR, we are planning to develop new abstractions, such as `FileSource` and `DataSource`. Would it be possible to design these abstractions in a way that allows both `opendal` and `object_store` to be used seamlessly with DataFusion? Similar to what we have done for Parquet's [`AsyncFileReader`](https://docs.rs/parquet/54.0.0/parquet/arrow/async_reader/trait.AsyncFileReader.html) and [`AsyncFileWriter`](https://docs.rs/parquet/54.0.0/parquet/arrow/async_writer/trait.AsyncFileWriter.html).

berkaysynnada replied:
Thank you for the suggestion, that could be a good idea. Is there any example having this abstraction to take reference, or available docs for OpenDAL API? IMO, this might be better addressed in a separate PR to keep the scope focused.

crepererum commented:
Could this migration be done step-by-step, instead of migrating ALL sources in a single monster PR with a +4k/-2k diff? How about we start w/ the traits and generic infrastructure, then migrate the in-mem source, then maybe parquet, then the next type, ... all in separate PRs?

ozankabak commented:
Thanks for all the review. We will focus on this again next week, improve the PR, and ask for more thoughts from the community.

mertak-synnada commented:
Here are some changes on plans,

Old ParquetExec usage:
```rust
use std::sync::Arc;
use arrow::datatypes::Schema;
use datafusion::datasource::physical_plan::{FileScanConfig, ParquetExec};
use datafusion::datasource::listing::PartitionedFile;
use datafusion_execution::object_store::ObjectStoreUrl;
use datafusion_physical_expr::expressions::lit;

let file_schema = Arc::new(Schema::empty());
let object_store_url = ObjectStoreUrl::local_filesystem();
let predicate = lit(true);
// Create a ParquetExec for reading `file1.parquet` with a file size of 100MB
let file_scan_config = FileScanConfig::new(object_store_url, file_schema)
   .with_file(PartitionedFile::new("file1.parquet", 100*1024*1024));
let exec = ParquetExec::builder(file_scan_config)
  // Provide a predicate for filtering row groups/pages
  .with_predicate(predicate)
  .build();
```

New ParquetSource usage:
```rust
use std::sync::Arc;
use arrow::datatypes::Schema;
use datafusion::datasource::physical_plan::FileScanConfig;
use datafusion::datasource::physical_plan::parquet::source::ParquetSource;
use datafusion::datasource::listing::PartitionedFile;
use datafusion_execution::object_store::ObjectStoreUrl;
use datafusion_physical_expr::expressions::lit;
use datafusion_physical_plan::source::DataSourceExec;
use datafusion_common::config::TableParquetOptions;

let file_schema = Arc::new(Schema::empty());
let object_store_url = ObjectStoreUrl::local_filesystem();
let predicate = lit(true);
let source = Arc::new(
    ParquetSource::new(
        Arc::clone(&file_schema),
        Some(predicate),
        None,
        TableParquetOptions::default()
    )
);
// Create a DataSourceExec for reading `file1.parquet` with a file size of 100MB
let file_scan_config = FileScanConfig::new(object_store_url, file_schema, source)
   .with_file(PartitionedFile::new("file1.parquet", 100*1024*1024));
let exec = file_scan_config.new_exec();
```

Old CsvExec usage:
```rust
let exec = CsvExec::builder(conf)
    .with_has_header(has_header)
    .with_delimeter(self.options.delimiter)
    .with_quote(self.options.quote)
    .with_terminator(self.options.terminator)
    .with_escape(self.options.escape)
    .with_comment(self.options.comment)
    .with_newlines_in_values(newlines_in_values)
    .with_file_compression_type(self.options.compression.into())
    .build();
Ok(Arc::new(exec))
```

New CsvExec usage:
```rust
let source = Arc::new(
    CsvSource::new(has_header, self.options.delimiter, self.options.quote)
        .with_escape(self.options.escape)
        .with_terminator(self.options.terminator)
        .with_comment(self.options.comment),
);
conf = conf.with_source(source);

Ok(conf.new_exec())
```

An example change on delta-rs for checking the source plan by type

From:
```rust
fn pre_visit(&mut self, plan: &dyn ExecutionPlan) -> Result<bool, Self::Error> {
    if let Some(parquet_exec) = plan.as_any().downcast_ref::<ParquetExec>() {
        self.predicate = parquet_exec.predicate().cloned();
        self.pruning_predicate = parquet_exec.pruning_predicate().cloned();
    }
    Ok(true)
}
```

To:
```rust
fn pre_visit(&mut self, plan: &dyn ExecutionPlan) -> Result<bool, Self::Error> {
    if let Some(parquet_exec) = plan.as_any().downcast_ref::<DataSourceExec>() {
        let source = parquet_exec.source();
        let file_scan = source.as_any().downcast_ref::<FileScanConfig>().unwrap();
        let file_source = file_scan.file_source();
        let parquet = file_source.as_any().downcast_ref::<ParquetSource>().unwrap();

        self.predicate = parquet.predicate().cloned();
        self.pruning_predicate = parquet.pruning_predicate().cloned();
    }
    Ok(true)
}
```

ozankabak commented:
Dear all, this PR is ready for the another round of reviews. I think it is in decent shape, and I'm looking forward to getting more feedback. Old source operators are merely wrappers of the new unified operator, and all tests pass. So we successfully avoided the two-impls issue (thanks @alamb for the suggestion).

Without any extra effort, we will already benefit from this architecture by having embedded projection capabilities for all data sources.

ozankabak reviewed (approved):
This looks good to me, and AFAICT all feedback has been addressed. No existing code is removed, we only have deprecations. We also took a look at how downstream projects would be affected in terms of migration (we looked at *delta-rs* for this purpose), and didn't see any big problems.

I am pinging all parties that showed interest (modulo my memory failing) below -- thank you all for the first round of reviews. The more eyes we can get on this before a merge, the better. If there are any more suggestions for improvements, we will be happy to address them as well.

datafusion/influxdata: @alamb
delta-rs: @ion-elgreco
comet: @andygrove, @parthchandra, @mbutrovich
synnada: @berkaysynnada, @jayzhan-synnada

also, @Xuanwo, @xudong963

alamb commented:
I will review this later today

@wiedld what do you think?

ion-elgreco commented:
Looks great! Migrating from the ParquetExec to DataSourceExec looks straightforward to me

alamb commented:
I am trying to upgrade InfluxDB IOx to use this PR. I will report back later today on its impact

alamb commented:
> (I am trying to upgrade InfluxDB IOx to use this PR. I will report back later today on its impact)

To be clear, while this is definitely selfish for us at Influx, I also think given how much of the DataFusion API we used, we are likely to hit many of the issues other users would hit, so I think testing our upgrade is a good way to test the impact of this change on all users

alamb reviewed (commented):
All in all the upgrade is currently going well.

I hit a few papercuts that would make it easier to upgrade -- but nothing a simple Upgrade Guide wouldn't solve

I am not done yet, but some stuff so far: (none of this is required in this PR in my opinion)

### How do I make a `FileSouce`?

`FileScanConfig::new()` now takes a different type, `FileSource`, but when I followed the docs to `FileSource`  it didn't tell me how how to create one, so I had to go digging around.

**Suggested action**: Add some doc hints / classes that implement FileSource such as a link to `ParquetSource`

I had to go look at https://github.com/apache/datafusion/pull/14224 to figure it out.

## constructing `ParquetFormat` was awkward

To make a `ParquetFormat` as I had to pass in several optional arguments, `predicate`, `metadata_size_hint`,

```rust
        let file_schema: Arc<Schema> = Arc::clone(&schema);
        let predicate: Option<Arc<dyn PhysicalExpr>> = None;
        let metadata_size_hint: Option<usize> = None;
        let table_parquet_options: TableParquetOptions = table_parquet_options();
        let file_source = Arc::new(ParquetSource::new(file_schema, predicate, metadata_size_hint, table_parquet_options));
        ```
**Suggested Action**: It would be nicer if new() took only required arguments and optional arguments had a `with_...` style API

```rust
  let file_schema: Arc<Schema> = Arc::clone(&schema);
  let table_parquet_options: TableParquetOptions = table_parquet_options();
  let file_source = Arc::new(ParquetSource::new(file_schema, table_parquet_options)
    .with_metadata_size_hint(...)
  );
```


## Where did `ParquetExecBuilder` go?


I had code like this"
```rust
        let builder = ParquetExecBuilder::new_with_options(base_config, table_parquet_options());
```

That I had to look to https://github.com/apache/datafusion/pull/14224  to figure out to change to
```rust
       conf.new_exec()
```

**Suggested Action**: Leave deprecated verson of ParquetExecBuilder with #[deprecated] and a hint about the new API

If the above is not possible, then a migration guide will likely help

ozankabak commented on datafusion/core/src/datasource/data_source.rs:
```diff
@@ -0,0 +1,264 @@
+    }
+}
+
+/// Determines file types
```
Let's create a ticket and address this with a quick follow-on to enable extensibility 🚀

ozankabak commented:
@alamb -- great, we can incorporate your three suggestions before merging. Keep them coming if anything else comes up. @mertak-synnada, let's address these (and anything else that may come up in the next 12 hours), and get this merge ready tomorrow 🚀 🚀 🚀

mertak-synnada commented:
@alamb The FileType enum was forgotten, I deleted it completely and returned just a &str for file_type()

Changed the ParquetSource initialization, now `new()` is only accepting `TableParquetOptions` , metadata_size_hint and predicate (also file_schema was related to it) moved into `with_xxx` functions.

Deprecated ParquetExecBuilder, and added link to ParquetSource example documentation. Also added a link from FileSource to ParquetSource and CsvSource examples.

Thanks for all the valuable feedback! I believe we're in a better state now 🚀

ozankabak commented:
> @alamb The FileType enum was forgotten, I deleted it completely and returned just a &str for file_type()

Great - custom source creators will just implement some `MyCoolFormatSource` just like `ParquetSource` implements `FileSource`. No need for hardcoded enums or another level of traits. They inform the overall mechanism of their file type extension by simply overriding the `file_type` method.

---

I think we addressed all the feedback and made the PR even better. Migration path is simple, and two major downstream projects took a pass on it. Let's wait for a few hours in case anything else comes in, and then proceed with the merge. @alamb, can you take a final look to verify all is good? Thanks.

alamb commented:
> @alamb, can you take a final look to verify all is good? Thanks.

Doing now -- I am figuring out the pattern for migration where we have code that looks for `ParquetExec` in the plan . Will have an answer in a few minutes.

alamb commented:
I found with this small change I could rewrite our optimizer passes fairly easily (it also seems to not change this particular PR). We can fix as a follow on PR

```diff
diff --git a/datafusion/physical-plan/src/source.rs b/datafusion/physical-plan/src/source.rs
index 0b7e6dd03..6f57870ce 100644
--- a/datafusion/physical-plan/src/source.rs
+++ b/datafusion/physical-plan/src/source.rs
@@ -176,8 +176,8 @@ impl DataSourceExec {

     /// Return the source object
     #[allow(unused)]
-    pub fn source(&self) -> Arc<dyn DataSource> {
-        Arc::clone(&self.source)
+    pub fn source(&self) -> &Arc<dyn DataSource> {
+        &self.source
     }
```



# Rationale
We have a bunch of code like this in influxdb_iox for rewriting ParquetExec in various ways:

https://github.com/influxdata/influxdb3_core/blob/a5f6076c966f4940a67998e0b85d12c3e8596715/iox_query/src/physical_optimizer/cached_parquet_data.rs#L52-L88

To make this easier I am trying to replicate the pattern of `plan.as_any().downcast_ref::<ParquetExec>` and I came up with this:

(However, the rust lifetime rules prevented this from compiling without the change above

```rust
/// A view of a [`DataSourceExec`] that scans parquet files
pub struct ParquetExecWrapper<'a> {
    base_config: &'a FileScanConfig,
    parquet_source: &'a ParquetSource,
}

impl<'a> ParquetExecWrapper<'a> {
    /// Create a new wrapper from a [`ExecutionPlan`], returning `None` if the
    /// plan is not a [`DataSourceExec`] that scans Parquet files
    pub fn try_new(plan: &'a dyn ExecutionPlan) -> Option<Self> {
        let plan = plan.as_ref();
        let data_source_exec = plan.as_any().downcast_ref::<DataSourceExec>()?;
        let base_config = data_source_exec
            .source()
            .as_any()
            .downcast_ref::<FileScanConfig>()?;
        let parquet_source = base_config
            .file_source()
            .as_any()
            .downcast_ref::<ParquetSource>()?;
        Some(Self {
            base_config,
            parquet_source,
        })
    }

    /// Return the underlying [`FileScanConfig`]
    pub fn base_config(&self) -> &'a FileScanConfig {
        self.base_config
    }

    /// Return the underlying [`ParquetSource`]
    pub fn parquet_source(&self) -> &'a ParquetSource {
        self.parquet_source
    }
}
```

Otherwise things are going well -- I am making good progress

ozankabak commented:
> Otherwise things are going well -- I am making good progress

Great! The change you suggested makes sense. Since it is a very small change, let's add to this PR directly and make migration smoother to others as well

alamb commented:
> Great! The change you suggested makes sense. Since it is a very small change, we'll shortly add it to this PR and so migration smoother to others as well.

To be clear, I think we can do these kinds of changes in a follow on PR too -- I haven't found any blockers to this PR being merged yet but I don't have our tests running yet to be sure. 

But all in all things are going well from my perspective

alamb reviewed (commented):
I got enough of our code to compile and tests running that i think this PR is ok to merge. Thank you @mertak-synnada @ozankabak and @berkaysynnada -- this is pretty epic

I believe this will be a non trivial change for any system that directly creates and manipulates `ParquetExec` but I think the new structures seem to support everything we need. 

It would be really great to write up a document (maybe a blog post) that explains this change and gives help for people upgrading. Specifically some examples of creating ParquetExec before and after (and something similar to `CsvExec` and `AvroExec`. 

As I work through the rest of our upgrade, I may have additional ideas to make it easier.

I can't realistically review this PR in detail, bug I trust that other committers have done so 👍

ozankabak commented:
> I got enough of our code to compile and tests running that i think this PR is ok to merge. Thank you @mertak-synnada @ozankabak and @berkaysynnada -- this is pretty epic

Great, let's go ahead 🚀 

> It would be really great to write up a document (maybe a blog post) that explains this change and gives help for people upgrading. Specifically some examples of creating ParquetExec before and after (and something similar to CsvExec and AvroExec.

This is a terrific idea -- we will be happy to write up a blog post that will serve a dual purpose: (1) help people with upgrading, and (2) brag about this epic PR 🙂

> As I work through the rest of our upgrade, I may have additional ideas to make it easier.

Great, we will be happy to collaborate on any follow-on PRs.

ozankabak commented:
I just merged the PR -- I will check back in half an hour to see if there are any problems

alamb commented:
> This is a terrific idea -- we will be happy to write up a blog post that will serve a dual purpose: (1) help people with upgrading, and (2) brag about this epic PR 🙂



Yes 100% that would be perfect. I think such a post would be find to put on the datafusion blog but content marketing wise if you write it putting it on a synnada blog would make total sense too

alamb commented:
Here is one small follow up from this PR: 
- https://github.com/apache/datafusion/pull/14670

alamb commented:
As I have been working through actually upgrading delta.rs with these changes it turns out it was more effort than expected (like everything in software)

I have created a few PRs that I think would be nice to get in before we release this change downstream
- https://github.com/apache/datafusion/pull/14882
- https://github.com/apache/datafusion/pull/14883
- https://github.com/apache/datafusion/pull/14898

alamb commented:
> It would be really great to write up a document (maybe a blog post) that explains this change and gives help for people upgrading. Specifically some examples of creating ParquetExec before and after (and something similar to CsvExec and AvroExec.



Also I started writing down an upgrade guide (not yet ready for review)
- https://github.com/apache/datafusion/pull/14891
